File: diag_frontend/src/api/userAnalytics.ts
import { fetchApi } from './fetch';

export type UserFeature =
  | 'diagnosis'
  | 'error_logs'
  | 'knowledge_base'
  | 'feedback'
  | 'user_analytics'
  | 'settings';

export interface UserAnalyticsSummary {
  total_users: number;
  new_users: number;
  active_users: number;
  today_active_users: number;
  total_usage: number;
  avg_usage_per_active_user: number;
  changes: {
    new_users: number;
    active_users: number;
    total_usage: number;
  };
}

export interface UserDailyUsage {
  date: string;
  new_users: number;
  active_users: number;
  usage_count: number;
}

export interface FeatureUsage {
  feature: string;
  count: number;
}

export interface UserUsageRow {
  id: string;
  name: string;
  itcode: string;
  email: string;
  created_at: string;
  last_login_at: string;
  last_active_at: string;
  login_count: number;
  diagnosis_count: number;
  usage_count: number;
  status: 'active' | 'dormant' | 'inactive';
}

export interface UserAnalyticsOverview {
  summary: UserAnalyticsSummary;
  daily: UserDailyUsage[];
  features: FeatureUsage[];
  users: {
    items: UserUsageRow[];
    total: number;
    page: number;
    limit: number;
  };
  generated_at: string;
}

function buildQuery(params: Record<string, string | number | undefined>): string {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') query.set(key, String(value));
  });
  return `?${query.toString()}`;
}

export const userAnalyticsApi = {
  getOverview(params: { days: number; page: number; limit?: number; search?: string }) {
    return fetchApi<UserAnalyticsOverview>(`/api/user-analytics/overview${buildQuery(params)}`);
  },

  trackFeature(feature: UserFeature) {
    return fetchApi<{ recorded: boolean }>('/api/user-analytics/events', {
      method: 'POST',
      body: JSON.stringify({ feature }),
    });
  },
};

File: diag_backend/app/routers/user_analytics.py
"""Authenticated user analytics endpoints."""

from typing import Literal

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel

from ..core.auth import get_current_user
from ..models.api import ApiResponse
from ..services.user_analytics_service import get_user_analytics_service

router = APIRouter(prefix="/user-analytics", tags=["用户分析"])


class UsageEventRequest(BaseModel):
    feature: Literal[
        "diagnosis",
        "error_logs",
        "knowledge_base",
        "feedback",
        "user_analytics",
        "settings",
    ]


@router.get("/overview", response_model=ApiResponse)
async def get_user_analytics_overview(
    days: int = Query(30, ge=7, le=90),
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=5, le=50),
    search: str | None = Query(None, max_length=100),
    current_user: dict = Depends(get_current_user),
):
    data = await get_user_analytics_service().get_overview(
        days=days,
        page=page,
        limit=limit,
        search=search,
    )
    return ApiResponse(success=True, data=data)


@router.post("/events", response_model=ApiResponse)
async def track_usage_event(
    request: UsageEventRequest,
    current_user: dict = Depends(get_current_user),
):
    await get_user_analytics_service().track_event(current_user, request.feature)
    return ApiResponse(success=True, data={"recorded": True})
